leakpro/attacks/utils/boosting.py

```diff
@@ -0,0 +1,212 @@
+        self.logger = logger
+        self.batch_size = batch_size
+
+    def run(self:Self) -> list:
+        """Run memorization enhancement."""
+
+        self.audit_data_length = len(self.audit_data_indices)
+        if self.org_audit_data_length*(1-self.memorization_threshold) < self.min_num_memorization_audit_points:
+            self.logger.info("Warning!")
+            self.logger.info("Memorization threshold gives less audit points than the minimum allowed")
+            self.logger.info('Please adjust "memorization_threshold" or "min_num_memorization_audit_points"')
+            self.memorization_threshold = 1-self.min_num_memorization_audit_points/self.org_audit_data_length
+            self.logger.info(f'Setting "memorization_threshold" to {self.memorization_threshold}')
+
+        if self.memorization_threshold != 0.0 and self.num_memorization_audit_points != 0:
+            self.logger.info("Warning!")
+            self.logger.info('"memorization_threshold" and "num_memorization_audit_points" is both used.')
+            self.logger.info('Defaulting to "num_memorization_audit_points"')
+
+        elif self.memorization_threshold != 0.0:
+            self.num_memorization_audit_points = int(self.org_audit_data_length * (1-self.memorization_threshold))
+
+        # Initialize memorization and privacy score
+        self.memorization_score = np.zeros(self.audit_data_length, dtype=float)
+        self.privacy_score = np.zeros(self.audit_data_length, dtype=float)
+
+        self._memorization_score()
+
+        if self.use_privacy_score:
+            self._privacy_score()
+
+        mem_mask, privacy_mask = self.adjust_memorization_mask()
+        mask = (mem_mask | privacy_mask)
+
+        return mask, self.memorization_score, self.privacy_score
+
+    def _memorization_score(self:Self) -> None:
+        """Run memorization score enhancement.
+
+        Memorization enhances the attack performance by only inlucing vulnerable data points
+        """
+
+        logits_function = ModelLogits()
+        logits = np.swapaxes(logits_function(self.shadow_models, self.handler, self.audit_data_indices, self.batch_size), 0, 1)
+        target_logits = np.swapaxes(logits_function([self.target_model], self.handler, self.audit_data_indices, self.batch_size),\
+                                    0, 1).squeeze()
+
+        logits = self.softmax_logits(logits)
+        target_logits = self.softmax_logits(target_logits)
+
+        if self.online:
+            for i, (logit, mask, label) in tqdm(enumerate(zip(logits, self.in_indices_mask, self.audit_data_labels)),
+                                                total=len(logits),
+                                                desc="Calculating memorization score",
+                                                leave=False
+                                                ):
+                self.memorization_score[i] = np.mean(logit[mask, label]) - np.mean(logit[~mask, label])
+        else:
```

Comment:
What do you think, does this description of the offline implementation make sense? @johanos1 
\# Offline impementation details.
            \# 1. Assumption made: The target logits will be a good approximate to the missing IN-samples
            \# 2. Given that memorization scores of the target model will likely be higher for IN-samples 
            \#       and low for OUT-samples, the offline version does more directly impact the 
            \#       filtering of IN- vs. OUT-samples.